### PR TITLE
Recipient profile top-5 updated

### PR DESCRIPTION
**Pull request by kojodoesdesign**
https://federal-spending-transparency.atlassian.net/browse/DEV-1234

- Added Top 5 Tables to Recipient Profile page
- Upgraded Code to es6 class implementation
- Text is truncated in the same way as state profile pages so that each table is a consistent height
- Added loading, error, and "No data available" states to tables
- Updated to match new endpoint
code review, design review, and backend merged

- [x] Code Review
- [x] Design Review
- [x] Backend Merged

### CODE REVIEW

File: api_contracts/contracts/search/AdvancedSearch.md [hunk 1]
```diff
@@ -29,6 +29,10 @@ This endpoint returns a list of the top results of specific categories sorted by
             + `naics`
             + `county`
             + `district`
+            + `federal_accounts`
+            + `product_service`
```
Comment (ebdabbs) [inline api_contracts/contracts/search/AdvancedSearch.md]: `product_service` is the same as `psc`. (Update in `topCategories.js` as well)

File: src/_scss/pages/recipient/recipientPage.scss [hunk 2]
```diff
@@ -1,7 +1,10 @@
 .usa-da-recipient-page {
 	@import "all";
 	@import "layouts/default/default";
-    @import "layouts/default/stickyHeader/header";
+  @import "layouts/default/stickyHeader/header";
+
+	@import "./topFive/topFive";
```
Comment (ebdabbs) [inline src/_scss/pages/recipient/recipientPage.scss]: Since the styling is mostly the same, you can import `"pages/state/topFive/topFive"` and override/ add styling as needed.

File: src/js/components/recipient/topFive/TopFive.jsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                             % of Total
+                            </th>
+                        </tr>
+                    </thead>
+                    <tbody
+                        className={`category-table__table-body ${hideBody}`}>
+                        {rows}
+                    </tbody>
+                </table>
+                <div className="category-table__link">
```
Comment (ebdabbs) [inline src/js/components/recipient/topFive/TopFive.jsx]: These links are out of scope for MVP and can be hidden/ removed for now.

File: src/js/components/recipient/topFive/TopFiveSection.jsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+        return (
+            <div
+                className="recipient-section topfive"
+                id="recipient-top-five">
+                <h3 className="state-section__title">
+                    Top 5
+                </h3>
+                <hr className="results-divider" />
+                <div className="state-section__description">
```
Comment (ebdabbs) [inline src/js/components/recipient/topFive/TopFiveSection.jsx]: Double check with the designers, but I believe we'll want to remove the last sentence for now since the links are out of scope.

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 5]
```diff
@@ -0,0 +1,166 @@
+/**
+ * TopFiveContainer.jsx
+ * Created by Kevin Li 5/15/18
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: Update comment here,` topCategories.js`, and `TopFiveContainer-test.jsx`

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        else if (this.props.fy !== 'all' && this.props.fy) {
+            const range = convertFYToDateRange(parseInt(this.props.fy, 10));
+            timePeriod = {
+                start_date: range[0],
+                end_date: range[1]
+            };
+        }
+
+        const filters = {
+            place_of_performance_scope: 'domestic',
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: We should be filtering on recipient instead of location here using `recipient_search_text` (which will search DUNS or recipient name)
See documentation for the standard filter object: https://github.com/fedspendingtransparency/usaspending-website/blob/dev/api_contracts/contracts/search/AdvancedSearch.md#filterobject-object

File: api_contracts/contracts/search/AdvancedSearch.md [hunk 7]
```diff
@@ -30,7 +30,6 @@ This endpoint returns a list of the top results of specific categories sorted by
             + `county`
             + `district`
             + `federal_accounts`
-            + `product_service`
             + `countries`
             <!-- + `states_and_territories` -->
```
Comment (ebdabbs) [inline api_contracts/contracts/search/AdvancedSearch.md]: We should work with backend to determine variable names for the new categories and then add them here.

File: api_contracts/contracts/recipient/RecipientProfile.md [hunk 8]
```diff
@@ -57,8 +57,10 @@ This endpoint returns a list of child recipients belonging to the given parent r
     The aggregate monetary value of all transactions associated with this recipient for the given time period.
 + total_transactions: 327721 (required, number)
     The number of transactions associated with this recipient for the given time period.
++ recipientHash: a1e2f7b4-073b-1174-255b-8a32d0d6a6dd-P (required, string)
```
Comment (ebdabbs) [inline api_contracts/contracts/recipient/RecipientProfile.md]: This change is taken care of in [this PR](https://github.com/fedspendingtransparency/usaspending-website/pull/838).

File: src/_scss/pages/state/topFive/_table.scss [hunk 9]
```diff
@@ -87,7 +90,32 @@
 
     .category-table__message {
         margin: rem(10) 0;
-        font-size: rem(12);
+				color:$color-gray-light;
+        font-size: rem(18);
         text-align: center;
+				background-color: #FCFCFC;
```
Comment (ebdabbs) [inline src/_scss/pages/state/topFive/_table.scss]: Could you ensure indentation is consistent through this file?

File: src/js/components/recipient/topFive/TopFive.jsx [hunk 10]
```diff
@@ -77,12 +77,14 @@ export default class TopFive extends React.Component {
                         {rows}
                     </tbody>
                 </table>
+
+                { /* Out of current sprint scope
```
Comment (ebdabbs) [inline src/js/components/recipient/topFive/TopFive.jsx]: I would either leave this in and hide it with CSS, or remove the commented out code for now.

File: src/js/models/v2/recipient/BaseRecipientOverview.js [hunk 11]
```diff
@@ -16,6 +16,7 @@ const BaseRecipientOverview = {
         this._totalTransactions = parseFloat(data.total_transactions) || 0;
         this.businessTypes = data.business_types || [];
         this.level = data.recipient_level || 'R';
+        this.recipientHash = data.recipientHash || null;
```
Comment (ebdabbs) [inline src/js/models/v2/recipient/BaseRecipientOverview.js]: Also updated in [this PR](https://github.com/fedspendingtransparency/usaspending-website/pull/838). Once it's merged, you can merge dev into your branch to test.

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 12]
```diff
@@ -64,12 +68,7 @@ export class TopFiveContainer extends React.Component {
         }
 
         const filters = {
-            place_of_performance_scope: 'domestic',
-            place_of_performance_locations: [
-                {
-                    country: 'USA'
-                }
-            ]
+            recipientHash: this.props.recipientHash
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: The filter param name should be `recipient_id`

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+        );
+    }
+}
+
+
+export default connect(
+    (state) => ({
+        total: state.recipient.overview._totalAmount,
+        fy: state.recipient.fy,
+        recipientHash: state.recipient.overview.recipientHash
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: When the recipient overview contract is updated, you will be able to access the hash via `state.recipient.id`

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+            });
+    }
+
+    parseResults(data, type) {
+        const parsed = data.map((item, index) => {
+            const result = Object.create(BaseStateCategoryResult);
+            result.populate(item, index + 1);
+
+            // use a special naming template for DUNS
+            if (type === 'recipient_duns') {
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: There is no DUNS table for the recipient profile, so this block can be removed. However, you will want to look at the mockup and set the naming template for anything that doesn't follow the default `code - name` (found in BaseStateCategoryResult).

Comment (ebdabbs) [reply src/js/containers/recipient/topFive/TopFiveContainer.jsx]: See https://github.com/fedspendingtransparency/usaspending-website/blob/dev/src/js/containers/state/topFive/TopFiveContainer.jsx#L151 for agency/ subagency name template

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    dataParams() {
+        let timePeriod = null;
+        if (this.props.fy === 'latest') {
+            const trailing = getTrailingTwelveMonths();
+            timePeriod = {
+                start_date: trailing[0],
+                end_date: trailing[1]
+            };
+        }
+        else if (this.props.fy !== 'all' && this.props.fy) {
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: We were previously excluding the time period filter for `all`, but after discussing with @dpb-bah, we will need to pass the beginning of FY 2008 through the end of the current FY in that case.

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 16]
```diff
@@ -59,6 +61,15 @@ export class TopFiveContainer extends React.Component {
                 end_date: trailing[1]
             };
         }
+        else if (this.props.fy === 'all') {
+            const currentFY = currentFiscalYear();
+            const earliestRange = convertFYToDateRange(parseInt(earliestFiscalYear, 10) + 1);
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: You don't need to add 1 here (Fiscal year 2008 starts Oct 1 2007)

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 17]
```diff
@@ -59,6 +61,15 @@ export class TopFiveContainer extends React.Component {
...
+            const LatestRange = convertFYToDateRange(parseInt(currentFY, 10));
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: Could you make this variable lowercase?

File: api_contracts/contracts/search/AdvancedSearch.md [hunk 18]
```diff
@@ -29,9 +29,9 @@ This endpoint returns a list of the top results of specific categories sorted by
             + `naics`
             + `county`
             + `district`
-            + `federal_accounts`
-            + `countries`
-            <!-- + `states_and_territories` -->
+            + `federal_account`
+            + `country`
+            <!-- + `state_territory` -->
```
Comment (ebdabbs) [inline api_contracts/contracts/search/AdvancedSearch.md]: Can this be uncommented?

File: src/js/components/recipient/topFive/TopFiveSection.jsx [hunk 19]
```diff
@@ -9,13 +9,38 @@ import { recipientCategories as topCategories } from 'dataMapping/recipients/top
 import TopFiveContainer from 'containers/recipient/topFive/TopFiveContainer';
 
 export default class TopFiveSection extends React.Component {
-    render() {
-        const content = topCategories.map((category) => (
-            <TopFiveContainer
+    chunk(array, size) {
```
Comment (ebdabbs) [inline src/js/components/recipient/topFive/TopFiveSection.jsx]: Move this logic to the render method and simplify/ improve readability (see DM)

File: api_contracts/contracts/search/AdvancedSearch.md (hunk 18 above)
Comment (kojodoesdesign) [reply api_contracts/contracts/search/AdvancedSearch.md]: Ah, I commented this out to test the error messages

File: src/js/containers/recipient/topFive/TopFiveContainer.jsx [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+        if (this.props.fy === 'latest') {
+            const trailing = getTrailingTwelveMonths();
+            timePeriod = {
+                start_date: trailing[0],
+                end_date: trailing[1]
+            };
+        }
+        else if (this.props.fy === 'all') {
+            const currentFY = currentFiscalYear();
+            const earliestRange = convertFYToDateRange(parseInt(earliestFiscalYear, 10) + 1);
```
Comment (ebdabbs) [inline src/js/containers/recipient/topFive/TopFiveContainer.jsx]: This should be `(parseInt(earliestFiscalYear, 10))` (Fiscal year 2008 starts Oct 1 2007)
